jaeger-core/src/main/java/com/uber/jaeger/samplers/RemoteControlledSampler.java

```diff
@@ -86,36 +88,31 @@ private void updateSampler() {
       return;
     }
 
-    Sampler newSampler;
-    try {
-      newSampler = extractSampler(response);
-    } catch (UnknownSamplingStrategyException e) {
-      metrics.samplerParsingFailure.inc(1);
-      return; // sampler updateGauge without a new sampler
-    }
+    Sampler oldSampler = sampler;
 
-    if (!this.sampler.equals(newSampler)) {
-      synchronized (this) {
-        this.sampler = newSampler;
-        metrics.samplerUpdated.inc(1);
+    if (response.getOperationSampling() != null) {
+      OperationSamplingParameters samplingParameters = response.getOperationSampling();
+      if (sampler instanceof PerOperationSampler) {
+        if (((PerOperationSampler) sampler).update(samplingParameters)) {
+          metrics.samplerUpdated.inc(1);
+        }
+      } else {
+        sampler = new PerOperationSampler(maxOperations, response.getOperationSampling());
       }
-    }
-  }
-
-  private Sampler extractSampler(SamplingStrategyResponse response)
-      throws UnknownSamplingStrategyException {
-    if (SamplingStrategyType.PROBABILISTIC.equals(response.getStrategyType())) {
+    } else if (response.getProbabilisticSampling() != null) {
       ProbabilisticSamplingStrategy strategy = response.getProbabilisticSampling();
-      return new ProbabilisticSampler(strategy.getSamplingRate());
-    }
-
-    if (SamplingStrategyType.RATE_LIMITING.equals(response.getStrategyType())) {
+      sampler = new ProbabilisticSampler(strategy.getSamplingRate());
+    } else if (response.getRateLimitingSampling() != null) {
       RateLimitingSamplingStrategy strategy = response.getRateLimitingSampling();
-      return new RateLimitingSampler(strategy.getMaxTracesPerSecond());
+      sampler = new RateLimitingSampler(strategy.getMaxTracesPerSecond());
+    } else {
+      metrics.samplerParsingFailure.inc(1);
+      log.info("No strategy present in response.");
     }
 
-    throw new UnknownSamplingStrategyException(
-        String.format("Unsupported sampling strategy type %s", response.getStrategyType()));
+    if (!oldSampler.equals(sampler)) {
```

Comment:
inefficient, especially for per-op sampler, since you already checked above it if's different or not.